src/CoercibleUtils.hs

```diff
@@ -1,8 +1,58 @@
-module CoercibleUtils where
+{- |
+Utility functions for Coercible types.
+Primarily pulled from the package `newtype-generics`, and
+based on Conor McBride's Epigram work, but
+generalised to work over anything `Coercible`.
 
-import Data.Coerce
+>>> ala Sum foldMap [1,2,3,4]
+10
+
+>>> ala Endo foldMap [(+1), (+2), (subtract 1), (*2)] 3
+8
+
+>>> under2 Min (<>) 2 1
+1
+
+>>> over All not (All False)
+All {getAll = True)
+
+Users might also find the GHC plugin <https://github.com/mpickering/hashtag-coerce hashtag-coerce> useful in tandem with this library.
+-}
+module CoercibleUtils
+  ( (#.), (.#)
+  , op
+  , ala, ala'
+  , under, over
+  , under2, over2
+  , underF, overF
+  ) where
+
+import Data.Coerce (Coercible, coerce)
 
 -- | Coercive left-composition.
+--
+--   The coercive composition operators solve a problem as presented in <https://ghc.haskell.org/trac/ghc/ticket/7542 GHC Trac #7542>.
+--
+--   The problem, in a nutshell:
+--
+--   If N is a newtype constructor, then N x will always have the same
+--   representation as x (something similar applies for a newtype deconstructor).
+--   However, if f is a function,
+--   
+--   N . f = \x -> N (f x)
+--   
+--   This looks almost the same as f, but the eta expansion lifts it--the lhs could
+--   be `_|_`, but the rhs never is. This can lead to very inefficient code. Thus we
+--   steal a technique from Shachaf and Edward Kmett and adapt it to the current
+--   (rather clean) setting. Instead of using N . f, we use N #. f, which is
+--   just
+--   
+--   coerce f \`asTypeOf\` (N . f)
+--   
+--   That is, we just pretend that f has the right type, and thanks to the safety
+--   of coerce, the type checker guarantees that nothing really goes wrong. We still
+--   have to be a bit careful, though: remember that #. completely ignores the
+--   value of its left operand."
```

Comment:
I think the background info here applies equally to both operators and should be in a separate [documentation chunk](http://haskell-haddock.readthedocs.io/en/latest/markup.html#named-chunks-of-documentation).

Regarding the warning that we ignore the value of one argument: This applies to all exports, so maybe we should put it into the module description.

Also the markup here could be better. But I can fix that myself.